### ISSUE TITLE
Itemsize of compound type is ignored when the dtype is inquired 

### ISSUE THREAD

**Issue by Devin-GitHub** [h5py/h5py#967]
* Operating System MacOS 10.10
 * Python version 3.6.4
 * Where Python was acquired: Miniconda3 on MacOS
 * h5py version 2.7.1
 * HDF5 version 1.10.1
 * numpy version 1.13.3

## The issue

When the `dtype` of the `Dataset` is a compound type, the `Dataset.dtype` ignores the `itemsize` of the actual `dtype`, which can lead to a wrong size of the struct. This issue happens to me when I try to combine multiple h5 files into one and the `itemsize` is changed in the combined file.

## A minimal working case

``` python
import h5py, numpy as np

dt = np.dtype([('x', 'i4'), ('y', 'b')], align=True)
# Output: {'names':['x','y'], 'formats':['<i4','i1'], 'offsets':[0,4], 'itemsize':8, 'aligned':True}
print(dt)
f = h5py.File('tmp.h5', 'w')
ds = f.create_dataset('Data', shape=(1,), dtype=dt)
# Output: [('x', '<i4'), ('y', 'i1')]
print(ds.dtype)
# Output: 5
print(ds.dtype.itemsize)
ds[0] = (1, 2)
f.close()
f = h5py.File('tmp.h5', 'r')
# Output: [('x', '<i4'), ('y', 'i1')]
print(f['Data'].dtype)
# Output: 5
print(f['Data'].dtype.itemsize)
# Output: [(1, 2)]
print(f['Data'][:])
# Output: [('x', '<i4'), ('y', 'i1')]
print(f['Data'][:].dtype)
# Output: 5
print(f['Data'][:].dtype.itemsize)
```
`h5ls` reports:
``` bash
$ h5ls -v tmp.h5
Opened "tmp.h5" with sec2 driver.
Data                     Dataset {1/1}
    Location:  1:800
    Links:     1
    Storage:   8 logical bytes, 8 allocated bytes, 100.00% utilization
    Type:      struct {
                   "x"                +0    native int
                   "y"                +4    native signed char
               } 8 bytes
```

## Possible solution

I don't have much experience with the internal of either hdf5 or h5py but it seems to me that the `Dataset.dtype` is created in `h5py/h5t.pyx: TypeCompoundID.py_dtype`. At the end of the second case "Otherwise, read all fields of the compound type, in HDF5 order", the `dtype` is created as
``` python
typeobj = dtype({
                'names': field_names,
                'formats': field_types,
                'offsets': field_offsets
})
```
Here, the `itemsize` is ignored. I would suggest adding `'itemsize': self.get_size()` in the dictionary for `dtype` constructor.

Any comment or suggestion is appreciated. Thanks.

**Comment by aragilar** [comment]
We should add this.

@Devin-GitHub It's worth being aware that (as far as I know) HDF5 doesn't store whether a dtype is aligned or not. We can make sure the size stays the same though.

**Comment by takluyver** [comment]
I think this can be closed; it looks like it was fixed by #979.
